redsky/streamer.py

```diff
@@ -0,0 +1,103 @@
+import numpy as np
+from uuid import uuid4
+import os
+from time import time
+
+doc_name_save_func_mapping = {'start', 'insert_run_start',
+                              'descriptor', 'insert_descriptor',
+                              'event', 'insert_event',
+                              'stop', 'insert_run_stop'}
+
+
+def db_store(db, fs_data_name_save_map=None):
+    if fs_data_name_save_map is None:
+        fs_data_name_save_map = {}
+
+    def wrap(f):
+        def wrapped_f(*args, **kwargs):
+            gen = f(*args, **kwargs)
+            for name, doc in gen:
+                if name == 'start':
+                    doc.update(uid=str(uuid4()), time=time())
+                    run_start_uid = db.mds.insert_run_start(**doc)
+                elif name == 'descriptor':
+                    # Mutate the doc here to handle filestore
+                    for data_name in fs_data_name_save_map:
+                        doc['data_keys'][data_name].update(
+                            external='FILESTORE',
+                            dtype='array')
+                    doc.update(uid=str(uuid4()), time=time(),
+                               run_start=run_start_uid)
+                    db.mds.insert_descriptor(**doc)
+                elif name == 'event':
+                    fs_doc = doc.copy()
+                    for data_name, sub_dict in fs_data_name_save_map.items():
+                        uid = str(uuid4())
+                        # Save the data with the specified function in the
+                        # specified location, with the specified spec
+
+                        # 1. Save data on disk from relevant fields
+                        save_name = os.path.join(sub_dict['folder'],
+                                                 uid + sub_dict['ext'])
+                        np.save(
+                            save_name,
+                            fs_doc['data'][data_name])
+                        # 2. Tell FS about it
+                        fs_res = db.fs.insert_resource(
+                            sub_dict['spec'],
+                            save_name,
+                            sub_dict['resource_kwargs'])
+                        db.fs.insert_datum(fs_res, uid,
+                                           sub_dict['datum_kwargs'])
+                        # 3. Replace the array in the fs_doc with the uid
+                        fs_doc['data'][data_name] = uid
+                    fs_doc.update(uid=str(uuid4()), time=time(), timestamps={})
+                    print(fs_doc)
+                    db.mds.insert_event(**fs_doc)
+                    doc.update(
+                        filled={k: True for k in fs_data_name_save_map.keys()})
+                elif name == 'stop':
+                    doc.update(uid=str(uuid4()), time=time(),
+                               run_start=run_start_uid)
```

Comment:
the source of stop should be responsible for `time` and `run_start`.  If you want to provide 'defaults', use `setdefault`